File: ts/src/123.ts
//309. Best Time to Buy and Sell Stock with Cooldown
function getAns(
  prices: number[],
  ind: number,
  buy: number,
  dp: number[][],
): number {
  const n = prices.length;
  if (ind >= n) return 0;

  if (dp[ind][buy] !== -1) return dp[ind][buy];

  let profit: number;

  if (buy === 0) {
    // We can buy the stock
    profit = Math.max(
      getAns(prices, ind + 1, 0, dp),
      -prices[ind] + getAns(prices, ind + 1, 1, dp),
    );
  } else {
    // We can sell the stock
    profit = Math.max(
      getAns(prices, ind + 1, 1, dp),
      prices[ind] + getAns(prices, ind + 2, 0, dp),
    );
  }

  return (dp[ind][buy] = profit);
}

function maxProfit(prices: number[]): number {
  const n = prices.length;
  const dp: number[][] = new Array(n)
    .fill(null)
    .map(() => new Array(2).fill(-1));

  return getAns(prices, 0, 0, dp);
}

// Example Usage:
console.log(maxProfit([1, 2, 3, 0, 2])); // Example test case

//sub: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/submissions/1552561637/
//sol: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/solutions/6458072/309-best-time-to-buy-and-sell-stock-with-1tlz/

File: rust/src/123.rs
//123. Best Time to Buy and Sell Stock III
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        if prices.is_empty() {
            return 0;
        }

        let mut first_buy = i32::MIN;
        let mut first_sell = 0;
        let mut second_buy = i32::MIN;
        let mut second_sell = 0;

        for price in prices {
            first_buy = first_buy.max(-price); // First buy
            first_sell = first_sell.max(first_buy + price); // First sell
            second_buy = second_buy.max(first_sell - price); // Second buy
            second_sell = second_sell.max(second_buy + price); // Second sell
        }

        second_sell // Return max profit after two transactions
    }
}
//sub: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/submissions/1580869873/?envType=study-plan-v2&envId=top-interview-150
//sol: https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/solutions/6561315/123-best-time-to-buy-and-sell-stock-iii-v5168/
